README.md
---
layout: home
title: "armonika - Encoding/decoding/handling of variable length numbers in bit addressable memory"
image: assets/favimage-840x472.jpg
---

## Work-In-Progress

# armonika

Encoding/decoding/handling of variable length numbers in bit addressable memory

Instead of an ad with tracking...  
Like to [donate ![opencollective](assets/opencollective-icon.svg)](https://opencollective.com/RockingShip) some appreciation for the use or inspiration this gives you?

## Welcome to the Wonderful World of self describing streaming numbers

*When everything is relative and you missed the origin*

variable length/serial bus vs. fixed width/parallel.

advantage of serial/small footprint with self terminiating encoding.

The terminator marks either that "0" or "1" would continue indefinitely.

3 components,
- the streaming i/o port
- simple arithmetic and logic.
- validation

implelemted as pull-event driven.

serial has the added advantage of a smaller footprint.

Encodes a delimiter in the data stream.
That is N-consecutive same values.
To allow that pattern to occur in data, the third bit

Encoding/decoding/handling of variable length numbers in bit addressable memory.

Addressing data in memory requires two components: location (address) and size (bit count).

With fixed width memory (multiple of power-of-2 bytes) the maximum storage size is determined at compile-time
whereas with variable-width memory the unrestricted storage size is determined at run-time.

This project is an experimental environment for storing data in bit-addressable memory.
It is intended to bridge the worlds of the `xtools` and `untangle` projects.

The proposed encoding has these features:
 - Streaming.
 - No maximum size.
 - Simple and fast encoding/decoding.
 - Simple and fast length detection.
 - Simple functional implementation of bitwise and integer arithmetic operators.
 - Simple hardware implementation.

@date 2020-07-04 02:28:54

This encoding gives two major benefits.
- There is no maximum to the amount of storage you want to address.
- Every value has an infinite number of leading "0" or "1"'s.
Overflows will never occur because storage and encoding is variable-width.
Being variable-width, strings no longer need a null terminator.
end-of-sequence is embedded in the encoding of the value.
The only thing needed to know is where the sequence ends, not how.
Encounding is immune for "Y2K-bug" and "8-bit/64-bit compatible" issues.

# Encoding/decoding scheme

When you lost track of how many bits there are in an int...

```
    After n consecutive "0", force emit (on encoding) a "1" which is (force ignored) during decoding.
    n+1 consecutive "0" indicitates leading-zeros/end-of-number.
```

`untangle` in unfamiliar with the concept "two-complement" whereas `xtools` assumed its presence.
Negative Two-complement numbers have an infinite number of leading "1".
Extend the encoding that:

```
    After n consecutive "1", force emit (on encoding) a "0" which is (force ignored) during decoding.
    n+1 consecutive "1" indicitates leading-ones/end-of-number.
```

# Examples

The following tables demonstrates the encoding with nthe numbers 0-19.

The binary input is read right-to-left, the string output is read left-to-right. 

| Input binary | Output N=2 | output N=3 |
|------:|:-----------|:-----------|
| ----< | >-------   | >-------
| 00000 | 000        | 0000      
| 00001 | 1000       | 10000     
| 00010 | 01000      | 010000    
| 00011 | 11000      | 110000    
| 00100 | 0011000    | 0010000   
| 00101 | 101000     | 1010000   
| 00110 | 011000     | 0110000   
| 00111 | 1101000    | 1110000   
| 01000 | 00101000   | 000110000 
| 01001 | 10011000   | 10010000  
| 01010 | 0101000    | 01010000  
| 01011 | 110011000  | 11010000  
| 01100 | 001101000  | 00110000  
| 01101 | 1011000    | 10110000  
| 01110 | 01101000   | 01110000  
| 01111 | 11011000   | 111010000 
| 10000 | 0010011000 | 0001010000
| 10001 | 100101000  | 1000110000
| 10010 | 010011000  | 010010000 
| 10011 | 1100101000 | 110010000 
   
The first number that does not fit in 32 bits:
 - for N=2. 43691 (0x0000aaab) which is encoded as "110011001100110011001100110011000".
 - for N=3. 629144 (0x00099998) which is encoded as "110011001100110011001100110011000".
 - for N=4. 2098063 (0x0020038f) which is encoded as "111100001111000010000100001100000".
 - for N=5. 2220512 (0x0021e1e0) which is encoded as "000001111100000111110000011000000".

For unsigned encoding only the number 15 is different:
    N=2, signed "11011000", unsigned "111010000"  
    N=2, signed "111010000", unsigned "11110000"  

# Unsigned

Although the encoding supports variable-length negative numbers, this functionality is not always used.
Most procedural languages with 32/64 bits wordsize all work in modulo 2^32 (or 64) numerical space.
"-1" is not "minus 1" which is negative, "-1" is actually "-1 modulo 2^32 (or 64)" which is unsigned and therefor positive.

An additional implementation is included for unsigned variable-length numbers.  
  
# Frequency count

Lowest `N` is simplest implementation, higher `N` is greater storage efficiency.
To aid in making a choice the following table contains frequency counts for encoded lengths with different settings of `N`.
Counted are all the values in range 0-65535.

| length | N = 2 | N = 3 | N = 4 | N = 5 | notes |
|-------:|------:|------:|------:|------:|:------|
|  3 |     1 |       |       |       |
|  4 |     1 |     1 |       |       |
|  5 |     2 |     1 |     1 |       |
|  6 |     2 |     2 |     1 |     1 |
|  7 |     4 |     4 |     2 |     1 |
|  8 |     6 |     6 |     4 |     2 |
|  9 |    10 |    12 |     8 |     4 |
| 10 |    16 |    22 |    14 |     8 |
| 11 |    26 |    40 |    28 |    16 |
| 12 |    42 |    74 |    54 |    30 |
| 13 |    68 |   136 |   104 |    60 |
| 14 |   110 |   250 |   200 |   118 |
| 15 |   178 |   460 |   386 |   232 |
| 16 |   288 |   846 |   744 |   456 | &lt;--- maximum length with binary encoding
| 17 |   466 |  1556 |  1434 |   896 |
| 18 |   754 |  2862 |  2764 |  1762 |
| 19 |  1220 |  5264 |  5328 |  3464 |
| 20 |  1972 |  9682 | 10270 |  6810 |
| 21 |  3162 | 14614 | 19796 | 13388 |
| 22 |  4924 | 15076 | 16940 | 26320 |
| 23 |  7176 |  9836 |  6384 | 10606 |
| 24 |  9370 |  3864 |  1022 |  1324 |
| 25 | 10540 |   842 |    52 |    38 |
| 26 |  9900 |    84 |       |       |
| 27 |  7570 |     2 |       |       |
| 28 |  4600 |       |       |       |
| 29 |  2160 |       |       |       |
| 30 |   754 |       |       |       |
| 31 |   184 |       |       |       |
| 32 |    28 |       |       |       |
| 33 |     2 |       |       |       |

Using `N=2` are clearly not efficient and using `N>2` does not show significant differences.

Taking `N=3` as default has an average of ~22 bits which is ~40% more storage than 16-bits binary encoded.

With unsigned encoding:

| length | N = 2 | N = 3 | N = 4 | N = 5 | notes |
|-------:|------:|------:|------:|------:|:------|
|  3 |     1 |       |       |       |
|  4 |     1 |     1 |       |       |
|  5 |     2 |     1 |     1 |       |
|  6 |     3 |     2 |     1 |     1 |
|  7 |     6 |     4 |     2 |     1 |
|  8 |    11 |     7 |     4 |     2 |
|  9 |    20 |    14 |     8 |     4 |
| 10 |    37 |    27 |    15 |     8 |
| 11 |    68 |    52 |    30 |    16 |
| 12 |   125 |   100 |    59 |    31 |
| 13 |   230 |   193 |   116 |    62 |
| 14 |   423 |   372 |   228 |   123 |
| 15 |   778 |   717 |   448 |   244 |
| 16 |  1431 |  1382 |   881 |   484 | &lt;--- maximum length with binary encoding
| 17 |  2632 |  2664 |  1732 |   960 |
| 18 |  4841 |  5135 |  3405 |  1904 |
| 19 |  8904 |  9898 |  6694 |  3777 |
| 20 | 13793 | 19079 | 13160 |  7492 |
| 21 | 15106 | 17263 | 25872 | 14861 |
| 22 | 10812 |  7178 | 11215 | 29478 |
| 23 |  4846 |  1351 |  1602 |  5816 |
| 24 |  1281 |    95 |    63 |   271 |
| 25 |   176 |     1 |       |     1 |
| 26 |     9 |       |       |       |

The average encoding length is somewhere between 20 and 22.
With `N=2` the encoding has possibly the simplest implementation with a high yield.

# Add/Subtract

@date 2020-07-03 01:39:28

Considering that add/subtract are the most basic arithmetic operator
The only moment that raises an issue is when the result of the subtract is negative.
Negative also reflects an active carry-out.

The core logic of add/subtract is identical.

Per bit:
 - emit = carryin ^ left ^ right;
 - carryout = carryin ? left | right : left & right;
 
For ADD the initial carry-in is "0"
For SUB the initial carry-in is "1" and right is bit-inverted.
Implies that all subtracts can be rewritten as additions.
With no subtract functionality being used, removed the use of an active carry-out.

## Source code

Grab one of the tarballs at [https://github.com/RockingShip/smile/releases](https://github.com/RockingShip/armonika/releases) or checkout the latest code:

```sh
  git clone https://github.com/RockingShip/armonika.git
```

# Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/RockingShip/armonika/tags).

# License

This project is licensed under the GNU General Public License v3 - see the [LICENSE.txt](LICENSE.txt) file for details


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

```
2021-03-26 19:11:36 Added: Donate button.
2021-03-23 12:37:21 Added: `moonwalk` theme.
2020-07-18 00:36:32 Added handful divide by constants to `div.c` to asses the inner workings.
2020-07-16 23:04:20 Added `div.c` to research streaming `DIV`.
2020-07-12 23:00:59 Upgraded `srun3` with memory port (sequential access) and `ALU` for core opcodes/instructions.
2020-07-10 23:35:46 Renamed `srun3.c` into `srun3.cc`.
```

## 2020-07-10 22:48:53 [Version 1.1.0]

Positive+Negative numbers indicated by polarity end-of-sequence marker.  
`ADD/SUB/LSL/LSR/AND/XOR/OR` operators.  
Full Streaming except for `LSL/LSR` right-hand-side.  

```
2020-07-08 23:01:30 Added `ADD/SUB/LSL/LSR`.
2020-07-07 22:40:54 Added `AND/XOR/OR`.
2020-07-04 22:39:55 Abandoning `urun2.c` in favor of `srun3.c`.
2020-07-01 22:40:54 Added `LSL/LSR/AND/XOR/OR`.
2020-07-01 12:57:42 Third implementation of `OR` in `urun2.c` using loops.
2020-06-30 01:19:40 Second implementation of `OR` in `urun2.c` using generated states.
2020-06-30 01:19:40 Initial implementation of `OR` in `urun2.c`.
2020-06-29 14:07:33 Added `ufrequency.c` for unsigned encoding.
                    Renamed `frequency.c` into `sfrequency.c` for signed encoding. 
```

## 2020-06-29 13:23:01 Version 1.0.0

Basic proof of concept.  
Signed variable-width numbers with configurable runlength terminator.  

```
2020-06-29 13:03:05 Added support for negative numbers.
2020-06-26 17:06:38 Fixed incomplete terminator.
2020-06-26 15:59:58 Added `frequency.c`.
2020-06-24 00:29:02 Initial commit.
```

[Unreleased]: https://github.com/RockingShip/armonika/compare/v1.1.0...HEAD
[Version 1.1.0]: https://github.com/RockingShip/armonika/compare/v1.0.0...v1.1.0


srun3.c
/*
 * srun3.c
 *
 * @date 2020-07-04 22:55:00
 *
 * Implementation of opcodes for "signed runlength-3" encoding.
 *
 * @date 2020-07-01 02:53:59
 * Previous implementation had the least number of conditionals/jumps. but the bitstream overhead. oo.
 *
 * @date 2020-07-01 14:41:14
 *
 * Note about shrink wrapping leading zeros of the result.
 * Swallowing long lengths of leading zero to emit them all again in case a future "1" appears kills streaming.
 * To make shrink-wrap semi-possible, save the bit position of the last emitted "1".
 * On return, rewind to last "1" and append end-of-sequence marker.
 * The rewind breaks streaming.
 *
 * @date 2020-07-01 21:41:14
 *
 * Test patterns. With `RUNN=3` the longest streak will be 4 bits long.
 * With 3 data paths, using `"3*4+1"` bit values should be sufficient to cover all situations and their harmonics.
 *
 * @date 2020-07-01 14:41:14
 *
 * Drop support of shrink-wrapping results (like in `srun2.c`).
 * - With signed end-of-sequence the code is much more complicated
 * - it rewinds the memory pointer which might break streaming.
 */

/*
 *	This file is part of Armonika,
 *	Encoding/decoding/handling of variable length numbers in bit addressable memory.
 *	Copyright (C) 2020, xyzzy@rockingship.org
 *
 *	This program is free software: you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <unistd.h>

// maximum runlength before escaping
#define RUNN 3

// timer tick
int tick = 0;

// what is value of next bit in the sequence
#define nextraw(MEM, POS) ({ unsigned _pos=(POS); ((MEM)[_pos>>3] >> (_pos&7)) & 1; })
// write next bit of the sequence. Old value is undefined and needs explicit set/clear.
#define emitraw(MEM, POS, BIT) ({ unsigned _pos=(POS); if (BIT) (MEM)[_pos>>3] |= 1<<(_pos&7); else (MEM)[_pos>>3] &= ~(1<<(_pos&7)); })

/**
 * Read next input bit, optionally swallowing any escaping markers
 *
 * @param {number} STATE - state of runlength counter
 * @param {number} BIT - read bit. NOTE: do not change, also last read bit
 * @param {number} MEM - bit-addressable memory
 * @param {number} POS - bit-position
 * @return {number} - read data-bit
 */
#define nextcooked(STATE, BIT, MEM, POS) do {			\
	if (STATE) {						\
		if (BIT == nextraw(MEM, POS++)) {		\
			/* same polarity */			\
			STATE <<= 1; /* bump counter */		\
								\
			/* test for escape bit */		\
			if (STATE & (1 << RUNN)) {		\
				/* opposite bit = escape, same bit = EOS */	\
				STATE = nextraw(MEM, POS++) ^ BIT;		\
			}					\
		} else {					\
			/* opposite polarity */			\
			BIT = BIT ^ 1; /* invert */		\
			STATE = 2; /* first bit of sequence */	\
		}						\
	}							\
} while (0)

/**
 * Emit the next output bit, injecting an escape after two consecutive bits.
 *
 * @param {number} STATE - state of runlength counter
 * @param {number} LAST - last emitted bit
 * @param {number} BIT - bit to emit
 * @param {number} MEM - bit-addressable memory
 * @param {number} POS - bit-position
 */
#define emitcooked(STATE, LAST, BIT, MEM, POS) do { \
	/* test if output streak needs to be escaped */		\
	if (STATE & (1 << RUNN)) {				\
		emitraw(MEM, POS++, LAST ^ 1); /* opposite polarity */	\
		STATE = 1; /* reset sequence */			\
	}							\
								\
	/* bit to emit */					\
	emitraw(MEM, POS++, BIT);				\
								\
	STATE = LAST ^ BIT ?					\
		2 /* switching polarity, first bit of sequence */	\
	      :							\
		STATE << 1; /* bump counter */ ;		\
								\
	/* remember last emitted bit */				\
	LAST = BIT;						\
} while (0)

/**
 * Emit end-of-sequence marker
 * Keep emitting leading bits until maximum run-length reached followed by terminator.
 *
 * @param {number} STATE - state of runlength counter
 * @param {number} LAST - last emitted bit
 * @param {number} BIT - polarity
 * @param {number} MEM - bit-addressable memory
 * @param {number} POS - bit-position
 */
#define emitEOS(STATE, LAST, BIT, MEM, POS) do {		\
	/* keep emitting until maximum run-length reached */	\
	while (!(STATE & (1 << RUNN)) || LAST != BIT)		\
		emitcooked(STATE, LAST, BIT, MEM, POS);		\
								\
	/* finalise end-of-sequence */				\
	emitraw(MEM, POS++, BIT);				\
} while (0)


/**
 * Encode signed value
 *
 * @param {string} pDest - bit-addressable memory (LSB emitted first).
 * @param {number} bitpos - bit-position
 * @param {number} num - signed value
 * @return {number} - length including terminator.
 */
unsigned encode(unsigned char *pDest, unsigned bitpos, int64_t num) {
	unsigned state = 1; // reset sequence
	unsigned last = 0; // initial value not important as sequence is reset

	// as long as there are input bits
	while (num != 0 && num != -1) {
		// extract next LSB from input
		unsigned b = num & 1;
		num >>= 1; // NOTE: arithmetic shift leaves sign-bit untouched

		// inject into output
		emitcooked(state, last, b, pDest, bitpos);
	}

	// end-of-sequence polarity
	num &= 1;

	/*
	 * Buildup leading bits until maximum run-length reached
	 * It is safe to use bit0 of `num` as polarity
	 */
	while (!(state & (1 << RUNN)) || last != num)
		emitcooked(state, last, num, pDest, bitpos);

	// finalise end-of-sequence
	emitraw(pDest, bitpos++, num);

	return bitpos;
}

/**
 * Encode value into runlength-N, return string and length.
 * NOTE: encoded number is unsigned
 *
 * @param {string} pSrc - bit-addressable memory (LSB emitted first).
 * @param {number} bitpos - bit-position
 * @return {int64_t} - The decoded value as variable length structure. For demonstration purpose assuming it will fit in less that 64 bits.
 */
int64_t decode(unsigned char *pSrc, unsigned bitpos) {
	unsigned state = 1; // reset sequence
	unsigned b = 0;
	int64_t num = 0; // number being decoded
	unsigned numlen = 0; // length of `num` in bits

	// The condition is a failsafe as the runN terminator is the end condition
	do {
		nextcooked(state, b, pSrc, bitpos);
		num |= (uint64_t) b << numlen;
		numlen++;
	} while (state);

	// fill upper bits of resulting fixed-width number with polarity of end-of-sequence
	num |= -((uint64_t) b << numlen);

	return num;
}

/**
 * @date 2020-07-08 18:56:29
 *
 * Streaming ADD
 *
 * @param pDst 	 bit-memory base of result
 * @param dstpos - bit position of result
 * @param pL - bit-memory base of left-hand-side
 * @param iL - bit position of left-hand-side
 * @param pR - bit-memory base of right-hand-side
 * @param iR - bit position of right-hand-side
 * @return dstpos + encoded-length
 */
unsigned ADD(unsigned char *pDst, unsigned dstpos, unsigned char *pL, unsigned iL, unsigned char *pR, unsigned iR) {

	// three pipelines, two for left/right operands one for the result.
	// two consecutive "0" is the trigger. Either a "0" to end the sequence or "1" to escape and continue.
	//
	unsigned lstate = 1, lbit = 0;
	unsigned rstate = 1, rbit = 0;
	unsigned estate = 1, ebit, elast = 0;
	unsigned carry = 0;

	do {
		// load next data bit of input pipelines
		nextcooked(lstate, lbit, pL, iL);
		nextcooked(rstate, rbit, pR, iR);

		// operator
		ebit = carry ^ lbit ^ rbit;
		carry = carry ? lbit | rbit : lbit & rbit;

		// emit operator result
		emitcooked(estate, elast, ebit, pDst, dstpos);
	} while (lstate || rstate);

	// operator on final polarity
	ebit = carry ^ lbit ^ rbit;

	// end-of-sequence
	emitEOS(estate, elast, ebit, pDst, dstpos);

	return dstpos;
}

/**
 * @date 2020-07-08 18:56:29
 *
 * Streaming SUB
 *
 * NOTE: identical to `ADD` except right-hand-side and carry are inverted
 *
 * @param pDst 	 bit-memory base of result
 * @param dstpos - bit position of result
 * @param pL - bit-memory base of left-hand-side
 * @param iL - bit position of left-hand-side
 * @param pR - bit-memory base of right-hand-side
 * @param iR - bit position of right-hand-side
 * @return dstpos + encoded-length
 */
unsigned SUB(unsigned char *pDst, unsigned dstpos, unsigned char *pL, unsigned iL, unsigned char *pR, unsigned iR) {

	// three pipelines, two for left/right operands one for the result.
	// two consecutive "0" is the trigger. Either a "0" to end the sequence or "1" to escape and continue.
	//
	unsigned lstate = 1, lbit = 0;
	unsigned rstate = 1, rbit = 0;
	unsigned estate = 1, ebit, elast = 0;
	unsigned carry = 1;

	do {
		// load next data bit of input pipelines
		nextcooked(lstate, lbit, pL, iL);
		nextcooked(rstate, rbit, pR, iR);

		// operator
		ebit = carry ^ lbit ^ rbit ^ 1;
		carry = carry ? lbit | (rbit ^ 1) : lbit & (rbit ^ 1);

		// emit operator result
		emitcooked(estate, elast, ebit, pDst, dstpos);
	} while (lstate || rstate);

	// operator on final polarity
	ebit = (carry ^ 1) ^ lbit ^ rbit;

	// end-of-sequence
	emitEOS(estate, elast, ebit, pDst, dstpos);

	return dstpos;
}

/**
 * @date 2020-07-01 22:52:29
 *
 * Logical shift left
 *
 * Left-hand-side is streaming
 * Right-hand-size is enumerated and large values can critically impact operations.
 *
 * @param pDst 	 bit-memory base of result
 * @param dstpos - bit position of result
 * @param pL - bit-memory base of left-hand-side
 * @param iL - bit position of left-hand-side
 * @param pR - bit-memory base of right-hand-side
 * @param iR - bit position of right-hand-side
 * @return dstpos + encoded-length
 */
unsigned LSL(unsigned char *pDst, unsigned dstpos, unsigned char *pL, unsigned iL, unsigned char *pR, unsigned iR) {

	unsigned lstate = 1, lbit;
	unsigned estate = 1, ebit, elast = 0;

	/*
	 * decode rval
	 */
	int rval;
	rval = decode(pR, iR);

	/*
	 * emit `rval` number of "0"
	 */
	while (rval > 0) {
		emitcooked(estate, elast, 0, pDst, dstpos);
		--rval;
	}

	/*
	 * Copy lval to output
	 */
	do {
		nextcooked(lstate, lbit, pL, iL);
		emitcooked(estate, elast, lbit, pDst, dstpos);
	} while (lstate);

	// end-of-sequence
	emitEOS(estate, elast, lbit, pDst, dstpos);

	return dstpos;
}

/**
 * @date 2020-07-08 13:54:49
 *
 * Logical shift right
 *
 * Left-hand-side is streaming
 * Right-hand-size is enumerated and large values can critically impact operations.
 *
 * @param pDst 	 bit-memory base of result
 * @param dstpos - bit position of result
 * @param pL - bit-memory base of left-hand-side
 * @param iL - bit position of left-hand-side
 * @param pR - bit-memory base of right-hand-side
 * @param iR - bit position of right-hand-side
 * @return dstpos + encoded-length
 */
unsigned LSR(unsigned char *pDst, unsigned dstpos, unsigned char *pL, unsigned iL, unsigned char *pR, unsigned iR) {

	unsigned lstate = 1, lbit;
	unsigned estate = 1, ebit, elast = 0;

	/*
	 * decode rval
	 */
	int rval;
	rval = decode(pR, iR);

	/*
	 * Copy lval to output skipping first `rval` bits
	 */
	do {
		nextcooked(lstate, lbit, pL, iL);
		if (--rval < 0)
			emitcooked(estate, elast, lbit, pDst, dstpos);
	} while (lstate);

	// end-of-sequence
	emitEOS(estate, elast, lbit, pDst, dstpos);

	return dstpos;
}


/**
 * @date 2020-07-08 01:15:35
 *
 * Streaming AND
 *
 * @param pDst 	 bit-memory base of result
 * @param dstpos - bit position of result
 * @param pL - bit-memory base of left-hand-side
 * @param iL - bit position of left-hand-side
 * @param pR - bit-memory base of right-hand-side
 * @param iR - bit position of right-hand-side
 * @return dstpos + encoded-length
 */
unsigned AND(unsigned char *pDst, unsigned dstpos, unsigned char *pL, unsigned iL, unsigned char *pR, unsigned iR) {

	// three pipelines, two for left/right operands one for the result.
	// two consecutive "0" is the trigger. Either a "0" to end the sequence or "1" to escape and continue.
	//
	unsigned lstate = 1, lbit = 0;
	unsigned rstate = 1, rbit = 0;
	unsigned estate = 1, ebit, elast = 0;

	do {
		// load next data bit of input pipelines
		nextcooked(lstate, lbit, pL, iL);
		nextcooked(rstate, rbit, pR, iR);

		// operator
		ebit = lbit & rbit;

		// emit operator result
		emitcooked(estate, elast, ebit, pDst, dstpos);
	} while (lstate || rstate);

	// operator on final polarity
	ebit = lbit & rbit;

	// end-of-sequence
	emitEOS(estate, elast, ebit, pDst, dstpos);

	return dstpos;
}

/**
 * @date 2020-07-08 01:16:13
 *
 * Streaming XOR
 *
 * @param pDst 	 bit-memory base of result
 * @param dstpos - bit position of result
 * @param pL - bit-memory base of left-hand-side
 * @param iL - bit position of left-hand-side
 * @param pR - bit-memory base of right-hand-side
 * @param iR - bit position of right-hand-side
 * @return dstpos + encoded-length
 */
unsigned XOR(unsigned char *pDst, unsigned dstpos, unsigned char *pL, unsigned iL, unsigned char *pR, unsigned iR) {

	// three pipelines, two for left/right operands one for the result.
	// two consecutive "0" is the trigger. Either a "0" to end the sequence or "1" to escape and continue.
	//
	unsigned lstate = 1, lbit = 0;
	unsigned rstate = 1, rbit = 0;
	unsigned estate = 1, ebit, elast = 0;

	do {
		// load next data bit of input pipelines
		nextcooked(lstate, lbit, pL, iL);
		nextcooked(rstate, rbit, pR, iR);

		// operator
		ebit = lbit ^ rbit;

		// emit operator result
		emitcooked(estate, elast, ebit, pDst, dstpos);
	} while (lstate || rstate);

	// operator on final polarity
	ebit = lbit ^ rbit;

	// end-of-sequence
	emitEOS(estate, elast, ebit, pDst, dstpos);

	return dstpos;
}

/**
 * @date 2020-07-01 22:52:29
 *
 * Streaming OR
 *
 * @param pDst 	 bit-memory base of result
 * @param dstpos - bit position of result
 * @param pL - bit-memory base of left-hand-side
 * @param iL - bit position of left-hand-side
 * @param pR - bit-memory base of right-hand-side
 * @param iR - bit position of right-hand-side
 * @return dstpos + encoded-length
 */
unsigned OR(unsigned char *pDst, unsigned dstpos, unsigned char *pL, unsigned iL, unsigned char *pR, unsigned iR) {

	// three pipelines, two for left/right operands one for the result.
	// two consecutive "0" is the trigger. Either a "0" to end the sequence or "1" to escape and continue.
	//
	unsigned lstate = 1, lbit = 0;
	unsigned rstate = 1, rbit = 0;
	unsigned estate = 1, ebit, elast = 0;

	do {
		// load next data bit of input pipelines
		nextcooked(lstate, lbit, pL, iL);
		nextcooked(rstate, rbit, pR, iR);

		// operator
		ebit = lbit | rbit;

		// emit operator result
		emitcooked(estate, elast, ebit, pDst, dstpos);
	} while (lstate || rstate);

	// operator on final polarity
	ebit = lbit | rbit;

	// end-of-sequence
	emitEOS(estate, elast, ebit, pDst, dstpos);

	return dstpos;
}

/*
 * @date 2020-07-08 00:33:53
 *
 * Signal handler for timer tick
 */
void sigAlarm(int sig) {
	tick++;
	alarm(1);
}

unsigned char mem[512];
unsigned pos;

int main() {
	int64_t num;
	int polarity;

	setlinebuf(stdout);
	signal(SIGALRM, sigAlarm);
	alarm(1);

	/*
	 * Display numbers for visual inspection
	 */
	if (0) {
		for (num = -16; num <= +16; num++) {
			// rewind memory
			pos = 0;

			// encode test value
			pos = encode(mem, pos, num);

			printf("%3d: ", num);

			// display encoded answer
			unsigned k;
			for (k = 0; k < pos; k++)
				putchar(nextraw(mem, k) ? '1' : '0');

			// decode value
			int64_t n;
			n = decode(mem, 0);
			printf(" [%ld]", n);

			putchar('\n');
		}
	}

	/*
	 * Test that `encode/decode` counter each other.
	 */
	for (num = -(1 << 13); num <= +(1 << 13); num++) {
		// rewind memory
		pos = 0;

		// encode test value
		pos = encode(mem, pos, num);

		// decode value
		int64_t n;
		n = decode(mem, 0);

		if (n != num) {
			fprintf(stderr, "encode/decode error. Expected=%ld Encountered=%ld\n", num, n);
			return 1;
		}
	}

	/*
	 * Test that the encoded value has smallest storage.
	 * This is done by filling memory with quasi random bits, decoding+encoding, testing the size of representation being less-equal to the original
	 */
	for (polarity = 0; polarity < 2; polarity++) {
		for (num = 0x0000; num <= 0xffff; num++) {
			// setup memory
			mem[0] = num;
			mem[1] = (num >> 8);
			if (polarity)
				mem[2] = mem[3] = 0xff;
			else
				mem[2] = mem[3] = 0x00;

			// which bit is not part of the leading polarity
			int lengthDecode = 0;
			for (lengthDecode = 31; lengthDecode >= 0; --lengthDecode) {
				if (((mem[lengthDecode / 8] >> (lengthDecode % 8)) & 1) != polarity)
					break;
			}

			// decode value
			int64_t n;
			n = decode(mem, 0);

			// encode it again after presetting all bits of destination
			if (polarity)
				mem[0] = mem[1] = 0xff;
			else
				mem[0] = mem[1] = 0x00;
			encode(mem, 0, n);

			// which bit is not part of the leading polarity
			int lengthEncode = 0;
			for (lengthEncode = 31; lengthEncode >= 0; --lengthEncode) {
				if (((mem[lengthEncode / 8] >> (lengthEncode % 8)) & 1) != polarity)
					break;
			}

			// encode may not be longer than what was decoded
			if (lengthEncode > lengthDecode) {
				fprintf(stderr, "decode/encode length error. mem=%02x.%02x.%02x.%02x num=%ld lengthDecode=%d lengthEncode=%d\n",
					mem[3], mem[2], ((unsigned) num >> 8) & 0xff, (unsigned) num & 0xff, n, lengthDecode, lengthEncode);
				return 1;
			}
		}
	}

	/*
	 * Test all the basic operators
	 */
	unsigned round;
	for (round = 0; round < 10; round++) {
		// display round
		// @formatter:off
		switch (round) {
		case 0: continue; fputs("MUL\n", stdout); break;
		case 1: continue; fputs("DIV\n", stdout); break;
		case 2: continue; fputs("MOD\n", stdout); break;
		case 3: fputs("ADD\n", stdout); break;
		case 4: fputs("SUB\n", stdout); break;
		case 5: fputs("LSL\n", stdout); break;
		case 6: fputs("LSR\n", stdout); break;
		case 7: fputs("AND\n", stdout); break;
		case 8: fputs("XOR\n", stdout); break;
		case 9: fputs("OR\n", stdout); break;
		}
		// @formatter:on

		int64_t lval, rval;
		int progress = 0;
		for (lval = -(1 << 12); lval <= +(1 << 12); lval++) {
			for (rval = -(1 << 12); rval <= +(1 << 12); rval++) {
				// ticker
				progress++;
				if (tick) {
					fprintf(stderr, "\r\e[K%.2f%%", progress * 100.0 / (4.0 * (1 << 12) * (1 << 12)));
					tick = 0;
				}

				// rewind memory
				pos = 0;

				// encode <left>
				unsigned iL = pos;
				pos = encode(mem, pos, lval);

				// encode <right>
				unsigned iR = pos;
				pos = encode(mem, pos, rval);

				// perform opcode and evaluate native
				unsigned iOpcode = pos;
				int64_t expected = 0;
				switch (round) {
				case 0:
//					pos = MUL(mem, pos, mem, iL, mem, iR);
					expected = lval + rval;
					break;
				case 1:
//					pos = DIV(mem, pos, mem, iL, mem, iR);
					expected = lval + rval;
					break;
				case 2:
//					pos = MOD(mem, pos, mem, iL, mem, iR);
					expected = lval + rval;
					break;
				case 3:
					pos = ADD(mem, pos, mem, iL, mem, iR);
					expected = lval + rval;
					break;
				case 4:
					pos = SUB(mem, pos, mem, iL, mem, iR);
					expected = lval - rval;
					break;
				case 5:
					if (rval < 0 || rval > 20)
						continue;
					pos = LSL(mem, pos, mem, iL, mem, iR);
					expected = lval << rval;
					break;
				case 6:
					if (rval < 0 || rval > 20)
						continue;
					pos = LSR(mem, pos, mem, iL, mem, iR);
					expected = lval >> rval;
					break;
				case 7:
					pos = AND(mem, pos, mem, iL, mem, iR);
					expected = lval & rval;
					break;
				case 8:
					pos = XOR(mem, pos, mem, iL, mem, iR);
					expected = lval ^ rval;
					break;
				case 9:
					pos = OR(mem, pos, mem, iL, mem, iR);
					expected = lval | rval;
					break;
				};

				// extract
				int64_t answer;
				answer = decode(mem, iOpcode);

				/*
				 * Compare
				 */
				if (answer != expected) {
					fprintf(stderr, "result error 0x%lx OPCODE 0x%lx. Expected=0x%lx Encountered 0x%lx\n", lval, rval, expected, answer);
					return 1;
				}

				if (0) {
					// encode answer to determine length
					unsigned iAnswer = pos;
					pos = encode(mem, pos, answer);

					// display encoded answer
					int k;
					for (k = iOpcode; k < iAnswer; k++)
						putchar(nextraw(mem, k) ? '1' : '0');
					putchar('\n');

				}
			}
		}
		fprintf(stderr, "\r\e[K");
	}

	return 0;
}